tests/scripts/thread-cert/test_srp_server_reboot_port.py

```diff
@@ -0,0 +1,125 @@
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+import ipaddress
+import logging
+import unittest
+
+import command
+import thread_cert
+
+# Test description:
+#   This test verifies when an SRP server reboots, it will listen to a UDP port
+#   that wasn't used in the last time.
+#
+# Topology:
+#
+#   CLIENT (leader) -- SERVER (router)
+#
+
+CLIENT = 1
+SERVER = 2
+
+
+class SrpAutoStartMode(thread_cert.TestCase):
+    USE_MESSAGE_FACTORY = False
+    SUPPORT_NCP = False
+
+    TOPOLOGY = {
+        CLIENT: {
+            'name': 'SRP_CLIENT',
+            'masterkey': '00112233445566778899aabbccddeeff',
+            'mode': 'rdn',
+        },
+        SERVER: {
+            'name': 'SRP_SERVER',
+            'masterkey': '00112233445566778899aabbccddeeff',
+            'mode': 'rdn',
+        },
+    }
+
+    def test(self):
+        client = self.nodes[CLIENT]
+        server = self.nodes[SERVER]
+
+        #
+        # 0. Start the server & client devices.
+        #
+
+        client.srp_server_set_enabled(False)
+        client.start()
+        self.simulator.go(5)
+        self.assertEqual(client.get_state(), 'leader')
+
+        server.srp_server_set_enabled(True)
+        server.start()
+        self.simulator.go(5)
+        self.assertEqual(server.get_state(), 'router')
+
+        #
+        # 1. Enable auto start mode on client and check that server is used.
+        #
+
+        self.assertEqual(client.srp_client_get_state(), 'Disabled')
+        client.srp_client_enable_auto_start_mode()
+        self.assertEqual(client.srp_client_get_auto_start_mode(), 'Enabled')
+        self.simulator.go(2)
+
+        self.assertEqual(client.srp_client_get_state(), 'Enabled')
+        self.assertTrue(server.has_ipaddr(client.srp_client_get_server_address()))
+
+        ports = [server.get_srp_server_port()]
+
+        # Repeat the following steps for 20 times
+        for i in range(25):
+            #
+            # 2. Disable server and check client is stopped/disabled.
+            #
+
+            old_port = server.get_srp_server_port()
+            server.srp_server_set_enabled(False)
+            server.reset()
+            server.start()
+            self.simulator.go(5)
+
+            #
+            # 3. Enable server and check client starts again. Verify that the
+            # server is using a different port.
+
+            server.srp_server_set_enabled(True)
+            self.simulator.go(5)
+            self.assertEqual(client.srp_client_get_state(), 'Enabled')
+            self.assertEqual(client.srp_client_get_server_address(), server.get_mleid())
+            self.assertNotEqual(old_port, server.get_srp_server_port())
```

Comment:
I think we also need to verify that the SRP client will re-register its SRP service to the SRP server after SRP server is reset, which is the real purpose of this change.